In [9]:
%store -r

# Digits After Pi

# Introduction

The objective of the study is a simple analysis aiming at studying a possible underlying link between combinations of numbers and how many digits they take to be found into Pi. The analysis will only consider combinations which can be found within the first **million digits of Pi** because that is the best free option found online which is also contained into a Python module (**math-pi**).

In [10]:
import math_pi

pi = math_pi.pi(1, 100)

print(pi)

3.1415926535897932384626433832795028841971693993751058209749445923078164062862089986280348253421170679


The first part will introduce the function developed to check a given combination of numbers. Then, the combination of numbers considered in the analysis are presented and ordered according to some criteria. Eventually, the results are presented and an interpretation (if possible) is provided.

# The Check Function

The function developed (*main_computation*) takes a string as an input and returns a dictionary with the following keys:
<pre>
{
    "Digits Checked": [int],    # Indicates the number of Pi checked before finding the combination requested
    "Pi Until": [str],          # Stores the whole series of digits of Pi until the combination requested
    "Not Found": True           # Boolean value indicating the result of the query
}
</pre>

Hereby, an example is provided.

In [11]:
from Check_Numbers_Pi.Dependency import main_computation

example = main_computation("208")

print(example)

{'Digits Checked': 76, 'Pi Until': '3141592653589793238462643383279502884197169399375105820974944592307816406286208', 'Not Found': False}


# The Combinations

All the combinations of a maximum of 5 digits that can be created between using the numbers between 0 and 9 are taken into account.

To generate all the possible combinations, *itertools* module can be used. The idea is to generate the combinations of 1, 2, 3, 4 and then 5 digits and then collect them in an array.

Hereby, an example of such application considering 2 digits is provided.

Notice that the combinations generated are also sorted numerically: every list of combinations, in fact, is also ordered from 0 to the highest number that is possible to generate with the amount of digits considered.

In [12]:
import itertools

set = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

# First the combinations are generated
combinations_of_2_digits = [p for p in itertools.product(set, repeat = 2)]

print(combinations_of_2_digits)

print ("\n")

print (type(combinations_of_2_digits[0]))       # Showing the type of the first element (but they are all the same)

[(0, 0), (0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (0, 6), (0, 7), (0, 8), (0, 9), (1, 0), (1, 1), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (1, 8), (1, 9), (2, 0), (2, 1), (2, 2), (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (2, 8), (2, 9), (3, 0), (3, 1), (3, 2), (3, 3), (3, 4), (3, 5), (3, 6), (3, 7), (3, 8), (3, 9), (4, 0), (4, 1), (4, 2), (4, 3), (4, 4), (4, 5), (4, 6), (4, 7), (4, 8), (4, 9), (5, 0), (5, 1), (5, 2), (5, 3), (5, 4), (5, 5), (5, 6), (5, 7), (5, 8), (5, 9), (6, 0), (6, 1), (6, 2), (6, 3), (6, 4), (6, 5), (6, 6), (6, 7), (6, 8), (6, 9), (7, 0), (7, 1), (7, 2), (7, 3), (7, 4), (7, 5), (7, 6), (7, 7), (7, 8), (7, 9), (8, 0), (8, 1), (8, 2), (8, 3), (8, 4), (8, 5), (8, 6), (8, 7), (8, 8), (8, 9), (9, 0), (9, 1), (9, 2), (9, 3), (9, 4), (9, 5), (9, 6), (9, 7), (9, 8), (9, 9)]


<class 'tuple'>


Since the *main_computation* function requires inputs to be strings, we cannot use the elements generated until we have first converted all its elements into strings while keeping its sorting.

In [13]:
void_string = ""

i = 0

for combination in combinations_of_2_digits:

    for item in combination:

        void_string = void_string + str(item)

    combinations_of_2_digits[i] = void_string

    i += 1

    void_string = ""

print(combinations_of_2_digits)

print ("\n")

print (type(combinations_of_2_digits[0]))       # Showing the final type of the first element (but they are all the same)

['00', '01', '02', '03', '04', '05', '06', '07', '08', '09', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31', '32', '33', '34', '35', '36', '37', '38', '39', '40', '41', '42', '43', '44', '45', '46', '47', '48', '49', '50', '51', '52', '53', '54', '55', '56', '57', '58', '59', '60', '61', '62', '63', '64', '65', '66', '67', '68', '69', '70', '71', '72', '73', '74', '75', '76', '77', '78', '79', '80', '81', '82', '83', '84', '85', '86', '87', '88', '89', '90', '91', '92', '93', '94', '95', '96', '97', '98', '99']


<class 'str'>


Let's now apply the same algorithm to get the lists of all combinations of 1, 2, 3, 4 and 5 digits.

In [14]:
import numpy
import itertools

set = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

combinations = []

void_string = ""

for digits in range (0, 6, 1):                  # First generating the tuples

    i = 0

    combinations.append([p for p in itertools.product(set, repeat = digits)])
    
    for combination in combinations[digits]:    # Then converting the elements in each tuple into strings

        for item in combination:

            void_string = void_string + str(item)

        combinations[digits][i] = void_string

        i += 1

        void_string = ""

    combinations[digits] = numpy.array(combinations[digits])

with numpy.printoptions(threshold = 10):
    print(combinations[1])
    print(combinations[2])
    print(combinations[3])
    print(combinations[4])
    print(combinations[5])

print("\n")

%store combinations

['0' '1' '2' '3' '4' '5' '6' '7' '8' '9']
['00' '01' '02' ... '97' '98' '99']
['000' '001' '002' ... '997' '998' '999']
['0000' '0001' '0002' ... '9997' '9998' '9999']
['00000' '00001' '00002' ... '99997' '99998' '99999']


Stored 'combinations' (list)


# First Checks

Before starting the analysis, we should first check our set of combinations.

We will only consider combinations of numbers belonging to the maximum number of digits whose combinations will prove to be fully found within the first million digits of Pi.

To check so, we can use the attribute "Not Found" from the *main_computation* function.

<span style="color:red">**WARNING**</span>: the computation may take several minutes (around 15 minutes).

In [15]:
import numpy
from Check_Numbers_Pi.Dependency import main_computation

combinations_not_found = []

for digits in range (1, 6, 1):

    print("Checking combinations of {} digits...\n".format(digits))

    for i in range (0, len(combinations[digits]), 1):

        print("Current Iteration: {Iteration} / {Total_Iterations}". format(Iteration = i, Total_Iterations = len(combinations[digits])), end = "\r", flush = True)

        if (main_computation(combinations[digits][i])["Not Found"] == True):

            combinations_not_found.append(str(combinations[digits][i]))

combinations_not_found = numpy.array(combinations_not_found)

print("The following combinations have not been found within the first million digits of Pi:")

print(combinations_not_found)

print ("\n")

%store combinations_not_found

Checking combinations of 1 digits...

Checking combinations of 2 digits...

Checking combinations of 3 digits...

Checking combinations of 4 digits...

Checking combinations of 5 digits...

The following combinations have not been found within the first million digits of Pi:
['14523' '17125' '22801' '33394' '36173' '39648' '40527' '96710']


Stored 'combinations_not_found' (ndarray)


After running the *main_computation* function through all the combinations, it has been found that:

- All the numbers of 1, 2, 3 or 4 digits can be found within the first million digits of Pi.

- The following combinations of 5 digits cannot be found within the first million digits of Pi: 14523, 17125, 22801, 33394, 36173, 39648, 40527, 96710.

A quick proof of this is provided hereby:

In [16]:
from tabulate import tabulate
from Check_Numbers_Pi.Dependency import main_computation

for digits in range (1, 5, 1):

    print("Checking combinations of {} digits...\n".format(digits))

    for i in range (0, len(combinations[digits]), 1):

        print("Current Iteration: {Iteration} / {Total_Iterations}". format(Iteration = i, Total_Iterations = len(combinations[digits])), end = "\r", flush = True)

        if (main_computation(combinations[digits][i])["Not Found"] == True):

            print ((combinations[digits][i]))   # Nothing is printed, so all the combinations have been found

not_found_table = []

for combination in combinations_not_found:
    
    not_found_table.append([combination, main_computation(combination)["Not Found"]])

print(tabulate(not_found_table, headers = ["Combination", "Not Found"], tablefmt = "github", numalign = "center", stralign = "center"))    # Always True, so all of these combinations have not been found

Checking combinations of 1 digits...

Checking combinations of 2 digits...

Checking combinations of 3 digits...

Checking combinations of 4 digits...

|  Combination  |  Not Found  |
|---------------|-------------|
|     14523     |    True     |
|     17125     |    True     |
|     22801     |    True     |
|     33394     |    True     |
|     36173     |    True     |
|     39648     |    True     |
|     40527     |    True     |
|     96710     |    True     |
